### PR TITLE
[MPS] Fix and refactor unary/binary ops with non-zero offset or non-contiguous output

### LINKED ISSUE #100764: [MPS] Unary ops yield wrong results if striding is different
### 🐛 Describe the bug

Consider the following code
```python
x=torch.arange(4.0, device='mps').reshape(2, 2)
y=torch.empty(2, 2, device='mps').t()
print(torch.neg(x, out=y))
```
it will print
```
tensor([[-0., -2.],
        [-1., -3.]], device='mps:0')
```
but should have printed
```
tensor([[-0., -1.],
        [-2., -3.]], device='mps:0')
```
I guess error stems from this check:
https://github.com/pytorch/pytorch/blob/c9593bc0e15b661dc90286eeaf503f31a3c9df78/aten/src/ATen/native/mps/operations/UnaryOps.mm#L84-L88

### Versions

2.0/nightly

cc @ezyang @gchanan @zou3519 @kulinseth @albanD @DenisVieriu97 @razarmehr @abhudev

### PR DESCRIPTION
**Pull request by qqaatw**
Fixes #100764

This PR fixes the unary ops implementation and refactors the binary ops implementation a bit.

For unary ops:
Previously we didn't take into account unary ops that have a non-contiguous/storage-offset output, causing an incorrect result (because the MPS graph kernel always writes the buffer contiguously). Therefore, this PR creates a temporary output tensor for the graph first and then copy the result back to the original output tensor. We currently do not have a better fix other than this I think.

For binary ops, see https://github.com/pytorch/pytorch/pull/97085#discussion_r1140999125

See the added test for repro.

### CODE REVIEW

File: test/test_mps.py [hunk 1]
```diff
@@ -2356,6 +2356,12 @@ def helper(shape):
             self.assertEqual(y, ref_y)
             self.assertEqual(x.grad, cpu_x.grad)
 
+            # test in-place with storage_offset
+            y = x.detach()[1].tanh_()
+            ref_y = cpu_x.detach()[1].tanh_()
+
+            self.assertEqual(y, ref_y)
+
```
Comment (qqaatw) [inline test/test_mps.py:2364]: This doesn't test non-contiguous cases but the LSTM (macOS 12) tests actually test them

**Comment by qqaatw** [comment]
@malfet can you take a look at this please?

**Comment by qqaatw** [comment]
Hello @alband @malfet, could you review this PR?

**Comment by qqaatw** [comment]
Hello @albanD @malfet, could you review this PR please?

**Review by malfet — APPROVED** [review]
Looks good to me, though in the long run we should:
 - Move more tests to opinfo rather than keep adding specific cases to `test_mps`
 - Check for non-zero-offsetted tensors at lower level
 - Can't we just create new MTLBuffer from another MTLBuffer with offset rather than do the copy?

**Pull request closed without merging**